plutus-use-cases/src/Language/PlutusTx/Coordination/Contracts/Future.hs

```diff
@@ -74,92 +88,252 @@ The current value of the underlying asset is determined by an oracle. See note
+            st = FutureState { knownOutputs = initMp, futureData = fd }
+        flip evalStateT st $ 
+            (liftBase (endpoint "start") >>= futureContract)
+            <|>
+            (liftBase (endpoint @() "initialise") >> writeInitialTx)
+
+{- Note [Initialising the futures contract]
+
+The margin accounts of both parties (long & short) are kept in the same 
+transaction output. To initialise the contract we need a transaction that
+locks the initial margins of both parties with the contract script, as is done 
+in 'writeInitialTx', which is called from the "initialise" endpoint of the 
+contract. In the test suite we simply call "initialise" from one of the 
+emulator's wallets. To make this more realistic we could use an escrow contract 
+that takes the initial margins and then locks them using the script.
+
+TODO: Implement escrow contract that pays to a script address, and use it here.
+
+-}
+
+data Role = Long | Short
+    deriving (Generic, Show, FromJSON, ToJSON)
+
+-- | The current state of the futures contract.
+data FutureState =
+    FutureState
+        { knownOutputs :: AddressMap
+        -- ^ The unspent outputs (usually exactly 1) of the contract instance.
+        , futureData   :: FutureData
+        -- ^ The current margin accounts
+        }
+
+type MonadFuture r m =
+    ( MonadState FutureState m
+    , MonadReader Future m
+    , MonadBase (Resumable (Eff r)) m)
+-- TODO: ResumableT ?
+
+-- | Spend all known outputs at the contract address using the 'FutureRedeemer'.
+--   Use 'redeemWith0' instead if the transaction does not have a pay-to-script 
+--   output.
+redeemWith
+    :: MonadFuture r m
+    => FutureRedeemer
+    -> m [TxIn]
+redeemWith redeemer = do
+    valScript <- asks validatorScript
+    let redScript = redeemerScript redeemer
+    gets (AM.spendScriptOutputs valScript redScript . knownOutputs)
+
+-- | Spend all known outputs at the contract address using the 'FutureRedeemer'.
+redeemWith0
+    :: MonadFuture r m
+    => FutureRedeemer
+    -> m [TxIn]
+redeemWith0 redeemer = do
+    valScript <- asks validatorScript
+    let redScript = redeemerScript0 redeemer
+    gets (AM.spendScriptOutputs valScript redScript . knownOutputs)
+
+-- | Lock some Ada in the contract using the 'FutureData' as data script.
+payIntoContract
+    :: MonadReader Future m
+    => FutureData
+    -> Ada
+    -> m TxOut
+payIntoContract fd vl = do
+    address <- contractAddress
+    let datScript = DataScript $ Ledger.lifted fd
+    pure $ LTx.scriptTxOut' (Ada.toValue vl) address datScript
+
 -- | Initialise the futures contract by paying the initial margin.
---
-initialise :: (
-    MonadError WalletAPIError m,
-    WalletAPI m)
+writeInitialTx
+    :: ( Member WriteTx r
+       , MonadFuture r m )
+    => m ()
+writeInitialTx = do
+    dt <- gets futureData
+    initialMargin <- asks futureInitialMargin
+    out <- payIntoContract dt (2 * initialMargin)
```

Comment:
Why is is 2x? I think this is probably about what the contract does, but perhaps worth a comment.

Reply:
The initial margin needs to be paid by each participant (this was actually a mistake in the original contract). I'll add a comment.